### PR TITLE
🐈 Task: Refactor ersilia clear command #1266

### LINKED ISSUE #1266: 🐈 Task: Refactor ersilia clear command
### Summary

Related to #1262 

Ersilia clear command really uninstalls everything and therefore we want to rename it to `uninstall` to more clearly reflect that. At present it does not remove docker images for models that have been fetched from DockerHub. In addition to fixing that, we want to extend it in a way that ersilia is able to uninstall itself from the user's system.

### Objective(s)

- [ ] Refactor `clear` to `uninstall` in the command's [definition](https://github.com/ersilia-os/ersilia/blob/master/ersilia/cli/commands/clear.py). This would typically involve refactoring everything from the class name (`Clearer` -> `Uninstaller`), function names, updating the help text for the command, etc.
- [ ] This command can potentially use `ersilia delete --all` (ref: #1265) to delete all models, including the dockerized models as well. 
- [ ] Remove the ersilia installation as well. This could be a bit tricky depending on how ersilia was installed.

For now, the first two objectives are straightforward to address - and we can discuss approaches and edge cases for the last objective pertaining to ersilia uninstall.

### Documentation

_No response_

### PR DESCRIPTION
**Pull request by KimFarida**
Thank you for taking your time to contribute to Ersilia, just a few checks before we proceed
- [x] Have you followed the guidelines in our [Contribution Guide](https://github.com/ersilia-os/ersilia/blob/master/CONTRIBUTING.md)
- [ ] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?

**Description**

Ersilia clear command really uninstalls everything and therefore we want to rename it to uninstall to more clearly reflect that. At present it does not remove docker images for models that have been fetched from DockerHub. In addition to fixing that, we want to extend it in a way that ersilia is able to uninstall itself from the user's system.

**Changes to be made**
This PR aims to extend the uninstallation process for Ersilia by renaming the existing` clear` command to `uninstall `and implementing a more comprehensive uninstallation workflow. 
The changes made include:

I first refactored the code by changing every instance of `clear `and `Clearer` to` uninstall `and `Uninstaller` to indicate that the command will uninstall Ersilia rather than just clearing data.

Then I updated the` SimpleDocker `class and added a method that that removes Docker images associated with Ersilia.

To further help with the uninstallation process, i included a deletion of the conda env "erisilia" and refactored the directory removal logic to use a list of directories `(EOS, BENTOML_PATH) `, I also included logging where appropriate.

I implemented logic to uninstall the Ersilia package if it was installed via pip
[output.log](https://github.com/user-attachments/files/17328113/output.log)

**Status**

_Replace this line with what you have done so far..._

**To do**

I am not sure I have fully covered all need to completely purge Ersilia so i plan to discuss with my mentor on the next step forward.

Closes #1266

### CODE REVIEW

**Comment by dzumii** [comment]
@KimFarida Its a good practice to attach log files instead of screenshots, screenshots can be harder to read sometimes

**Comment by KimFarida** [comment]
> @KimFarida Its a good practice to attach log files instead of screenshots, screenshots can be harder to read sometimes

Ohhh. Thank you so much for the feedback. Ill add shortly. Meanwhile i would appreciate if you can review my PR, thank you again

**Comment by Malikbadmus** [comment]
Hi @KimFarida , Great job on this work! I have a few minor cleanup suggestions that I'd like to address before proceeding with the testing.

File: ersilia/utils/uninstall.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            logger.info(f"Removing Ersilia directory: {self.ersilia_dir}...")
+            try:
+                shutil.rmtree(self.ersilia_dir) 
+                logger.info("Ersilia directory removed successfully.")
+            except Exception as e:
+                logger.error(f"Failed to remove Ersilia directory: {e}")
+        else:
+            logger.warning("Ersilia directory not found; skipping removal.")
+
+    def _uninstall_ersilia_package(self):
```
[the remark this answers is not in the document]
Comment (KimFarida) [reply ersilia/utils/uninstall.py:29]: Yes, the logic was a bit wonky at the time and I wasn't sure how to proceed so i took it out haha. I'll make all the corrections. Thank you for the feedback!🙌

**Comment by KimFarida** [comment]
> Hi @KimFarida , Great job on this work! I have a few minor cleanup suggestions that I'd like to address before proceeding with the testing.

Thank you so much. Will implement the changes you suggested. In the  meantime may i know these changes?

File: ersilia/utils/docker.py [hunk 2]
```diff
@@ -320,7 +320,26 @@ def container_peak(self, model_id):
         except Exception as e:
             print(f"An error occurred: {e}")
```
Comment (DhanshreeA) [inline ersilia/utils/docker.py:321]: Can we also remove this print statement, please? Let's replace it with a debug log.

File: ersilia/utils/docker.py [hunk 3]
```diff
@@ -320,7 +320,26 @@ def container_peak(self, model_id):
...
+    def cleanup_ersilia_images(self):
+        """Remove all Ersilia-related Docker images"""
+        if self._with_udocker:
+            logger.warning("Docker cleanup not supported with udocker")
+            return
 
+        try:
+            images_dict = self.images()
+            
+            for image_name, image_id in images_dict.items():
```
Comment (DhanshreeA) [inline ersilia/utils/docker.py:333]: Where is this `images_dict` coming from?

**Review by DhanshreeA — CHANGES_REQUESTED** [review]

File: ersilia/utils/docker.py (hunk 2 above)
Comment (KimFarida) [reply ersilia/utils/docker.py:321]: Definitely!
I'll get back on changes in a couple of hours

File: ersilia/utils/clear.py [hunk 4]
```diff
@@ -24,6 +24,7 @@ def _conda(self):
             if len(env.split("-")[0]) == 7:
                 sc.delete(env)
 
-    def clear(self):
+    def uninstall(self):
         self._conda()
         self._directories()
+
```
[the remark this answers is not in the document]
Comment (DhanshreeA) [reply ersilia/utils/clear.py:30]: +1 to this

File: ersilia/utils/uninstall.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (KimFarida) [reply ersilia/utils/uninstall.py:29]: > There are no calls/references to this method within the Uninstaller class or any of the provided methods you defined here.

I just checked, its not in the most current change

File: ersilia/utils/clear.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (KimFarida) [reply ersilia/utils/clear.py:30]: clear.py?
Its not in the recent changes.

**Comment by KimFarida** [comment]
> @KimFarida Its a good practice to attach log files instead of screenshots, screenshots can be harder to read sometimes

Done!

File: ersilia/utils/docker.py (hunk 2 above)
Comment (KimFarida) [reply ersilia/utils/docker.py]: @DhanshreeA I'm done implementing changes @DhanshreeA

File: ersilia/utils/clear.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (KimFarida) [reply ersilia/utils/clear.py]: @Malikbadmus made the necessary changes

**Comment by Malikbadmus** [comment]
@KimFarida , Really great work!!

@DhanshreeA , I've thoroughly tested the code, and it functions as expected. LGTM!

**Comment by KimFarida** [comment]
> @KimFarida , Really great work!!
> 
> @DhanshreeA , I've thoroughly tested the code, and it functions as expected. LGTM!

Thank youuu!

File: ersilia/cli/commands/uninstall.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+def uninstall_cmd():
+    """Uninstalls all Ersilia artifacts present locally on the user's system"""
+
+    # Example usage: ersilia setup
+    @ersilia_cli.command(
+        short_help="Uninstall ersilia",
+        help="Uninstalls all Ersilia artifacts present locally on the user's system.",
+    )
+    def uninstall():
+        cl = Uninstaller()
```
Comment (DhanshreeA) [inline ersilia/cli/commands/uninstall.py:14]: We can rename this variable to `ui` instead of `cl`.

File: ersilia/utils/docker.py [hunk 6]
```diff
@@ -312,15 +312,38 @@ def container_peak(self, model_id):
                 return
 
         except docker.errors.NotFound as e:
-            print(f"Container {container.name} not found: {e}")
+            logger.debug(f"Container {container.name} not found: {e}")
```
Comment (DhanshreeA) [inline ersilia/utils/docker.py:315]: Can we make the logger an instance variable?

File: ersilia/utils/uninstall.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (DhanshreeA) [reply ersilia/utils/uninstall.py:15]: Have we tested if this actually works?

**Review by DhanshreeA — CHANGES_REQUESTED** [review]

**Comment by DhanshreeA** [comment]
Thanks @KimFarida happy to merge this soon!

File: ersilia/utils/docker.py (hunk 6 above)
Comment (KimFarida) [reply ersilia/utils/docker.py]: Done!

**Review by DhanshreeA — APPROVED** [review]

**Pull request merged**